File: 11-enums/main.go
package main

import "fmt"

// === Go 没有真正的枚举，使用 const + iota 模拟 ===

type Direction int

const (
	North Direction = iota
	South
	East
	West
)

// 为"枚举"实现 String() 方法
func (d Direction) String() string {
	switch d {
	case North:
		return "North"
	case South:
		return "South"
	case East:
		return "East"
	case West:
		return "West"
	default:
		return "Unknown"
	}
}

// === 带自定义值的"枚举" ===
type HttpStatus int

const (
	StatusOK            HttpStatus = 200
	StatusNotFound      HttpStatus = 404
	StatusInternalError HttpStatus = 500
)

func (s HttpStatus) String() string {
	switch s {
	case StatusOK:
		return "OK"
	case StatusNotFound:
		return "Not Found"
	case StatusInternalError:
		return "Internal Server Error"
	default:
		return "Unknown"
	}
}

// === 带方法的"枚举" ===
type Season int

const (
	Spring Season = iota
	Summer
	Autumn
	Winter
)

func (s Season) IsWarm() bool {
	return s == Spring || s == Summer
}

func (s Season) Next() Season {
	return (s + 1) % 4
}

func (s Season) Name() string {
	names := [...]string{"Spring", "Summer", "Autumn", "Winter"}
	if int(s) < len(names) {
		return names[s]
	}
	return "Unknown"
}

// === 用字符串枚举 ===
type Color string

const (
	Red   Color = "red"
	Green Color = "green"
	Blue  Color = "blue"
)

func main() {
	// === 基本使用 ===
	dir := North
	fmt.Println("方向:", dir)

	// === switch（不强制穷尽） ===
	switch dir {
	case North:
		fmt.Println("方向描述: 向北")
	case South:
		fmt.Println("方向描述: 向南")
	case East:
		fmt.Println("方向描述: 向东")
	case West:
		fmt.Println("方向描述: 向西")
	}

	// === 枚举方法 ===
	season := Summer
	fmt.Println("季节:", season.Name())
	fmt.Println("温暖:", season.IsWarm())
	fmt.Println("下一个:", season.Next().Name())

	// === 类型安全的弱点 ===
	// 任何 int 都能赋值给 Direction（类型别名）
	var invalid Direction = 99 // 编译通过！
	fmt.Println("无效方向:", invalid)

	// === HTTP 状态码 ===
	status := StatusNotFound
	fmt.Printf("HTTP 状态码: %d (%s)\n", int(status), status)

	// === 字符串枚举 ===
	color := Red
	fmt.Println("颜色:", color)

	// === iota 高级用法 ===
	type FilePermission int
	const (
		Read    FilePermission = 1 << iota // 1
		Write                              // 2
		Execute                            // 4
	)
	perms := Read | Write
	fmt.Printf("权限: %d (读=%v, 写=%v, 执行=%v)\n",
		perms,
		perms&Read != 0,
		perms&Write != 0,
		perms&Execute != 0,
	)

	// === 季节温度映射 ===
	temps := map[Season]int{
		Spring: 20,
		Summer: 35,
		Autumn: 15,
		Winter: -5,
	}
	fmt.Println("\n季节温度:")
	for s := Spring; s <= Winter; s++ {
		fmt.Printf("  %s: %d°C\n", s.Name(), temps[s])
	}
}

File: 11-enums/main.rs
use std::fmt;

// === 基本枚举 ===
#[derive(Debug, Clone, Copy, PartialEq)]
enum Direction {
    North,
    South,
    East,
    West,
}

impl fmt::Display for Direction {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Direction::North => write!(f, "North"),
            Direction::South => write!(f, "South"),
            Direction::East => write!(f, "East"),
            Direction::West => write!(f, "West"),
        }
    }
}

// === 指定底层类型 ===
#[derive(Debug, Clone, Copy)]
#[repr(u16)]
enum HttpStatus {
    Ok = 200,
    NotFound = 404,
    InternalError = 500,
}

// === 带方法的枚举 ===
#[derive(Debug, Clone, Copy, PartialEq)]
enum Season {
    Spring,
    Summer,
    Autumn,
    Winter,
}

impl Season {
    fn is_warm(&self) -> bool {
        matches!(self, Season::Spring | Season::Summer)
    }

    fn next(&self) -> Season {
        match self {
            Season::Spring => Season::Summer,
            Season::Summer => Season::Autumn,
            Season::Autumn => Season::Winter,
            Season::Winter => Season::Spring,
        }
    }

    fn name(&self) -> &str {
        match self {
            Season::Spring => "Spring",
            Season::Summer => "Summer",
            Season::Autumn => "Autumn",
            Season::Winter => "Winter",
        }
    }
}

// === Rust 枚举的真正威力：关联数据 ===
#[derive(Debug)]
enum Shape {
    Circle(f64),                      // 关联一个 f64
    Rectangle { width: f64, height: f64 }, // 关联命名字段
    Triangle { base: f64, height: f64 },
}

impl Shape {
    fn area(&self) -> f64 {
        match self {
            Shape::Circle(r) => std::f64::consts::PI * r * r,
            Shape::Rectangle { width, height } => width * height,
            Shape::Triangle { base, height } => 0.5 * base * height,
        }
    }

    fn name(&self) -> &str {
        match self {
            Shape::Circle(_) => "Circle",
            Shape::Rectangle { .. } => "Rectangle",
            Shape::Triangle { .. } => "Triangle",
        }
    }
}

// === Option 和 Result 就是枚举 ===
// enum Option<T> { Some(T), None }
// enum Result<T, E> { Ok(T), Err(E) }

fn divide(a: f64, b: f64) -> Result<f64, String> {
    if b == 0.0 {
        Err(String::from("division by zero"))
    } else {
        Ok(a / b)
    }
}

fn main() {
    // === 基本使用 ===
    let dir = Direction::North;
    println!("方向: {}", dir);

    // === match 穷尽性（必须覆盖所有变体） ===
    let msg = match dir {
        Direction::North => "向北",
        Direction::South => "向南",
        Direction::East => "向东",
        Direction::West => "向西",
    };
    println!("方向描述: {}", msg);

    // === 枚举方法 ===
    let season = Season::Summer;
    println!("季节: {}", season.name());
    println!("温暖: {}", season.is_warm());
    println!("下一个: {}", season.next().name());

    // === 枚举与整数转换 ===
    let status = HttpStatus::NotFound;
    let code = status as u16;
    println!("HTTP 状态码: {}", code);

    // === 关联数据（Rust 独有的强大功能） ===
    let shapes = vec![
        Shape::Circle(5.0),
        Shape::Rectangle { width: 4.0, height: 6.0 },
        Shape::Triangle { base: 3.0, height: 8.0 },
    ];

    println!("\n形状面积:");
    for shape in &shapes {
        println!("  {}: {:.2}", shape.name(), shape.area());
    }

    // === if let 语法糖 ===
    let maybe_value: Option<i32> = Some(42);
    if let Some(v) = maybe_value {
        println!("\n有值: {}", v);
    }

    // === Result 处理 ===
    match divide(10.0, 3.0) {
        Ok(result) => println!("10 / 3 = {:.2}", result),
        Err(e) => println!("错误: {}", e),
    }

    match divide(10.0, 0.0) {
        Ok(result) => println!("10 / 0 = {:.2}", result),
        Err(e) => println!("错误: {}", e),
    }

    // === 迭代所有枚举值（需要手动实现或使用 strum crate） ===
    let seasons = [Season::Spring, Season::Summer, Season::Autumn, Season::Winter];
    println!("\n季节温度:");
    let temps = [20, 35, 15, -5];
    for (season, temp) in seasons.iter().zip(temps.iter()) {
        println!("  {}: {}°C", season.name(), temp);
    }
}
